### PR TITLE
[rl] Add TITO generator and gen metrics

### PR DESCRIPTION
**Pull request by felipemello1**
This PR does three things:
- Changes generator to receive tokens instead of text (token-in/token-out) by encoding prompts in the controller and passing `tokenized_prompts` into the generator. This is the standard in RL training and avoids retokenization of texts, which leads to all sorts of issues.
- Adds generation_metrics to the generation output.
- Sets wandb=True in our configs

### test
image

[actor=<root>] ============================================================
[actor=<root>] Validation summary (pre / post):
[actor=<root>]   validation/reward/_max:  +1.300  /  +1.300
[actor=<root>]   validation/reward/_mean:  +0.385  /  +0.700
[actor=<root>]   validation/reward/_min:  +0.000  /  +0.300
[actor=<root>]   validation/reward/_std:  +0.541  /  +0.490
[actor=<root>]   validation/reward/_sum:  +7.700  /  +14.000
[actor=<root>]   validation/reward/component/correctness/mean:  +0.250  /  +0.400
[actor=<root>]   validation/reward/component/format/mean:  +0.135  /  +0.300
[actor=<root>] ============================================================

image

### CODE REVIEW

File: torchtitan/experiments/rl/config_registry.py [hunk 1]
```diff
@@ -23,6 +23,7 @@
 from torchtitan.experiments.rl.actors.generator import SamplingConfig, VLLMGenerator
 from torchtitan.experiments.rl.actors.trainer import PolicyTrainer
 from torchtitan.experiments.rl.grpo import GRPOLoss, RLTrainer
+from torchtitan.experiments.rl.observability import metrics as m
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/config_registry.py]: Here maybe import `MetricsProcessor` is better, as there's no sl / other m.metrics

File: torchtitan/experiments/rl/actors/generator.py [hunk 2]
```diff
@@ -364,30 +431,50 @@ async def generate(
...
 
             completions: list[Completion] = []
+            generation_metrics: list[m.Metric] = []
+            output_token_counts: list[int] = []
             for output in all_outputs:
                 prompt_idx = int(output.request_id)
-                prompt_token_ids = output.prompt_token_ids
+                # TODO: When renderer-owned prompt construction lands, check
+                # if we should return the output prompt token IDs
+                # or keep the input prompt token IDs
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/actors/generator.py]: when would they be different?

File: torchtitan/experiments/rl/actors/generator.py [hunk 3]
```diff
@@ -32,12 +33,71 @@
...
+
+    For `[num_prompts]` submitted prompts, vLLM returns `[num_prompts]`
+    per parent `RequestOutput`s (one per `add_request` call), each carrying
+    a single `RequestStateStats` on `.metrics`.
+
+    Caveat under `SamplingParams.n > 1`: vLLM stores one `RequestStateStats`
+    per child request; the parent output exposes the **last-finishing**
+    child's timeline. `arrival_time` is shared across siblings, but
+    [`queued_ts`, `scheduled_ts`, `first_token_ts`, `last_token_ts`,
+    `num_generation_tokens`] describe one specific child — not an aggregate,
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/actors/generator.py:56]: how about `num_cached_tokens`

File: torchtitan/experiments/rl/actors/generator.py [hunk 4]
```diff
@@ -32,12 +33,71 @@
...
+def _prepare_generation_request_metrics(
+    output: RequestOutput, *, prefix: str
+) -> list[m.Metric]:
+    """Prepare vLLM metrics from a RequestOutput.
+
+    For `[num_prompts]` submitted prompts, vLLM returns `[num_prompts]`
+    per parent `RequestOutput`s (one per `add_request` call), each carrying
+    a single `RequestStateStats` on `.metrics`.
+
+    Caveat under `SamplingParams.n > 1`: vLLM stores one `RequestStateStats`
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/actors/generator.py:52]: Given this "caveat", I assume the metrics is "lossy". Do we plan to fix / workaround it, e.g. by `SchedulerStats / IterationStats`?

**Review by tianyu-l — APPROVED** [review]
LGTM, had some questions.

File: torchtitan/experiments/rl/actors/generator.py (hunk 3 above)
Comment (felipemello1) [reply torchtitan/experiments/rl/actors/generator.py:56]: No, the cached_tokens is correct, the prompt is prefilled once and shared across siblings. 

That said, i will be modifying this so instead of doing n=8, we would do 8 generations of n=1, which is necessary for multiturn and also works for single turn. We will have the "caveat" fixed by default.

File: torchtitan/experiments/rl/actors/generator.py (hunk 4 above)
Comment (felipemello1) [reply torchtitan/experiments/rl/actors/generator.py:52]: answered above

File: torchtitan/experiments/rl/actors/generator.py (hunk 3 above)
Comment (tianyu-l) [reply torchtitan/experiments/rl/actors/generator.py:56]: hmm, I actually don't know how it works.
- In general, "prefix caching" shouldn't work only for prompt tokens? So I'd imagine that with prefix caching on, more things than prefix caching could be cached.
- When we have multiple requests, if they are all put in prefill stage together, would the cache be used at all?
- Is there other tradeoff between sending n=8 vs. sending 8 n=1 requests? Intuitively the former might be more efficient?

File: torchtitan/experiments/rl/actors/generator.py [hunk 5]
```diff
@@ -316,21 +377,25 @@ async def sync_log_step(self, step: int, relative_step: int | None = None) -> No
     @sl.log_trace_span("generate")
     async def generate(
         self,
-        prompts: list[str],
+        tokenized_prompts: list[list[int]],
         *,
         sampling_config: SamplingConfig | None = None,
-    ) -> list[Completion]:
-        """Generate completions for the given prompts.
+        metrics_prefix: str = "generator",
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/actors/generator.py:383]: what's the consideration of not hardcoding `is_validation` as in https://github.com/pytorch/torchtitan/blob/main/torchtitan/experiments/rl/observability/metrics/processor.py#L162

Comment (felipemello1) [reply torchtitan/experiments/rl/actors/generator.py:383]: People may have multiple generators, e.g. 2 models talking to each other, teacher/student, llm as a judge, etc. All sorts of ideas in RL around generators.

Comment (tianyu-l) [reply torchtitan/experiments/rl/actors/generator.py:383]: btw should we update the default here https://github.com/pytorch/torchtitan/blob/main/torchtitan/experiments/rl/observability/metrics/processor.py#L54

File: torchtitan/experiments/rl/grpo.py [hunk 6]
```diff
@@ -306,6 +307,8 @@ def __init__(self, config: Config):
             log_dir=config.dump_folder,
             job_config=config.to_dict(),
         )
+        # TODO: Replace this single-turn tokenizer with renderer
+        self.tokenizer = HuggingFaceTokenizer(tokenizer_path=config.hf_assets_path)
```
Comment (wwwjn) [inline torchtitan/experiments/rl/grpo.py:319]: In core titan, the tokenization is done during the data loader. Should we consider introduce a dataloader for RL as well (in the following PRs)?

File: torchtitan/experiments/rl/actors/generator.py (hunk 3 above)
Comment (wwwjn) [reply torchtitan/experiments/rl/actors/generator.py:56]: I have read the similar behavior of sending n 1-request to vllm before: https://fburl.com/code/ztfbbbew

Sending 8 n=1 requests helps better manage the unique ids in vllm requests (needs to dig into vllm for more details). However, these no information about the impact on perf

Comment (felipemello1) [reply torchtitan/experiments/rl/actors/generator.py:56]: I did a bit of digging:

1. Internally, vllm still breaks down the requests into single ones (per link that @wwwjn sent), and appends them to a queue. The scheduler batches multiple queued requests into a forward pass (continuous batching).
	
2. There's a scheduler that loops over requests and checks if their hash is a cache hit or miss. The first miss schedules the full prefill (e.g., 100 prompt tokens); the others register a cache hit — a reservation on the prefiller's blocks (the KV values themselves aren't computed yet). All 8 then run in one forward pass: within that pass, the prefiller writes the prompt KV, and the cache-hit siblings wait for it and read when ready (the kernel writes K, V before the attention step, so there's no race). After this forward, all 8 start decoding.
	
3. So if we do 8×n=1 or 1×n=8, we still calculate the full prompt prefill only once. One nice thing: vLLM does chunked prefill by default — new incoming prefill tokens get mixed into the same forward pass as ongoing decode tokens, so a new request doesn't stall in-flight ones. There's also prefill-decode disaggregation, which goes further by running prefill and decode on separate machines (each phase has a different compute/memory profile).
	
4. Prefix caching isn't prompt-only — any full block (default=16 tokens) of finalized tokens (prompt or generated) gets hashed and cached.

File: torchtitan/experiments/rl/actors/generator.py (hunk 5 above)
Comment (felipemello1) [reply torchtitan/experiments/rl/actors/generator.py:383]: i dont think that we want any of the generator metrics in the console log. We already have tokens/s. Let me know if you disagree

File: torchtitan/experiments/rl/grpo.py (hunk 6 above)
Comment (felipemello1) [reply torchtitan/experiments/rl/grpo.py:319]: Hmmm, we do have a `dataset` and a `make_env` that associates a sample to an env. I dont think we need a dataloader because there is no batching. For each dataset sample, we launch an async rollout request.

File: torchtitan/experiments/rl/actors/generator.py (hunk 2 above)
Comment (felipemello1) [reply torchtitan/experiments/rl/actors/generator.py]: They shouldnt be (I tested it). I got rid of it -- I removed it from "Completions". Now the `Trajectory` will just hold the `prompt_ids`, instead of the `Completion.prompt_ids`. We do: "token prompts in, token response out".

Having the generator output prompt tokens only makes sense if the input is text.
